File: case_studies/clift/matvec_matmul/matvec_matmul.cpp
#include <optitrust.h>

void matvec(int col_count, int red_count, float *x, float *y, float *w) {
  for (int j = 0; j < col_count; j++) {
    x[MINDEX1(col_count, j)] = 0.f;
    for (int k = 0; k < red_count; k++) {
      x[MINDEX1(col_count, j)] +=
          y[MINDEX1(red_count, k)] * w[MINDEX2(col_count, red_count, j, k)];
    }
  }
}
void iter_matvec(int row_count, int col_count, int red_count, float *const x,
                 float *const y, float *const w) {
  for (int i = 0; i < row_count; i++) {
    matvec(col_count, red_count, &x[MINDEX2(row_count, col_count, i, 0)],
           &y[MINDEX2(row_count, red_count, i, 0)], w);
  }
}
void matmul(int row_count, int col_count, int red_count, float *const x,
            float *const y, float *const w) {
  for (int i = 0; i < row_count; i++) {
    for (int j = 0; j < col_count; j++) {
      y[MINDEX2(row_count, col_count, i, j)] = 0.f;
      for (int k = 0; k < red_count; k++) {
        y[MINDEX2(row_count, col_count, i, j)] +=
            x[MINDEX2(row_count, red_count, i, k)] *
            w[MINDEX2(col_count, red_count, j, k)];
      }
    }
  }
}

File: case_studies/clift/matvec_matmul/matvec_matmul.ml
open Optitrust
open Prelude

let _ =
  Flags.pretty_matrix_notation := true;
  Flags.print_optitrust_syntax := true

let f = cFunDef "iter_matvec"

let _ =
  Run.script_cpp (fun () ->
      !!Function.inline [ cCall "matvec" ];
      !!Matrix.simpl_access_of_access ~indepth:true [ f ];

      !!Matrix.simpl_index_add [ nbMulti; f; cCellAccess ~base:[ cVar "x" ] (); cBinop Binop_add ];
      !!Matrix.simpl_index_add [ nbMulti; f; cCellAccess ~base:[ cVar "y" ] (); cBinop Binop_add ];

      !!Rewrite.equiv_at "int j; ==> 0 + j == j" [ nbMulti; cFunDef " iter_matvec" ] ~indepth:true;
      !!Function.uninline ~f:[ cFunDef "matmul" ] [ cFunBody "iter_matvec"; dSeqNth 0 ]
    )
